feat(2911): Support pipeline metrics in templateFactory

nathom opened the pull request:
## Context

It has been requested that template maintainers are able to see who is using which version, so that once
a new version has been released, those using an old version can be notified.

## Objective

Show the Pipelines that use a template version.

## References

https://github.com/screwdriver-cd/screwdriver/issues/2911

## License

<!-- The following line must be included in your pull request -->

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

nathom committed:
Add pipeline IDs and count to template

nathom committed:
Merge remote-tracking branch 'models/master'

jithine commented on lib/templateFactory.js:
```diff
@@ -91,11 +95,11 @@ class TemplateFactory extends BaseFactory {
+                        name: parsedTemplate.name,
                     },
-                    paginate
+                    paginate,
```
I don't think we use the style of comma after last field in a object

tkyi commented on lib/templateFactory.js:
```diff
@@ -104,18 +108,19 @@ class TemplateFactory extends BaseFactory {
+                paginate,
             })
-            .then(namespaceExists => {
+            .then((namespaceExists) => {
```
the added parens will fail lint

nathom replied:
What formatter do you guys use?

adong replied:
eslint + prettier

https://github.com/screwdriver-cd/eslint-config-screwdriver

adong replied:
```
node ./node_modules/.bin/eslint --fix .
```

You can run the above code, warnings are fine, but any errors will fail

nathom committed:
Add prettierrc

nathom committed:
Update formatting

nathom committed:
Fix indent

nathom committed:
Fix line width

nathom committed:
Add getWithMetrics method

nathom committed:
Switch to full name lookup from id for template metrics

nathom committed:
Handle case where >1 template used in pipeline

nathom committed:
Switch to pipelineUsage specification

nathom committed:
Fix build

nathom committed:
Merge branch 'screwdriver-cd:master' into master

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -273,24 +287,33 @@ class TemplateFactory extends BaseFactory {
+                    const buildCount = bCount.find(b => b.templateId === t.id);
+                    // The number of pipelines that use this template,
+                    // NOT the number of template uses in pipelines
+                    const pipelineCount = pCount.filter(p => p.templateId === t.id).length;
```
Tested against SQLite DB and noticed below issue

Items in `pCount` do not have `templateId`. So the `pipelineCount` is always 0.

<img width="482" alt="image" src="https://github.com/screwdriver-cd/models/assets/2124590/5e90967a-9d26-4df1-98f0-a885ee161467">

```
[
  {
    "pipelineId": 1,
    "count": 2
  },
  {
    "pipelineId": 2,
    "count": 2
  },
  {
    "pipelineId": 3,
    "count": 2
  },
  {
    "pipelineId": 4,
    "count": 3
  },
  {
    "pipelineId": 5,
    "count": 3
  },
  {
    "pipelineId": 6,
    "count": 3
  }
]
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -273,24 +287,33 @@ class TemplateFactory extends BaseFactory {
+            return Promise.all([
+                jobFactory.list(metricJobsListConfig),
+                buildFactory.list(metricBuildsListConfig),
+                jobFactory.list(metricPipelinesConfig)
```
Tested against SQLite DB and noticed below issues

1. The DB query results includes pipelines that do not have any job using template
2. It includes the pipelines with jobs using any template. Instead it should only include pipelines having jobs using the only the template specified in the config

**Query result:**
```
[
  {
    "pipelineId": 1,
    "count": 2
  },
  {
    "pipelineId": 2,
    "count": 2
  },
  {
    "pipelineId": 3,
    "count": 2
  },
  {
    "pipelineId": 4,
    "count": 3
  },
  {
    "pipelineId": 5,
    "count": 3
  },
  {
    "pipelineId": 6,
    "count": 3
  }
]
```

**Actual Jobs (Pipelines) using template:**
Only pipelines `4`, `5` and `6` have jobs using templates.
```
select pipelineId, id, name, templateId
from jobs
order by pipelineId;
```
```
+----------+--+-----------------------------+----------+
|pipelineId|id|name                         |templateId|
+----------+--+-----------------------------+----------+
|1         |1 |main                         |NULL      |
|1         |2 |publish                      |NULL      |
|2         |3 |validate                     |NULL      |
|2         |4 |publish                      |NULL      |
|3         |5 |validate                     |NULL      |
|3         |6 |publish                      |NULL      |
|4         |7 |job_not_using_template       |NULL      |
|4         |8 |job_using_template_apple_main|3         |
|4         |9 |job_using_template_apple_test|3         |
|5         |10|job_not_using_template       |NULL      |
|5         |11|job_using_template_pear_main |4         |
|5         |12|job_using_template_pear_test |4         |
|6         |13|job_not_using_template       |NULL      |
|6         |14|job_using_template_apple     |3         |
|6         |15|job_using_template_pear      |4         |
+----------+--+-----------------------------+----------+

```

sagar1312 replied:
Also encountered below error when tested against MySQL DB
```
230817/142821.021, (1692282500866:C02ZM198LVDV:491:llf9ab7u:10002) [request,server,error] data: Error
    at Query.run (/Users/dsagar/Workspace/screwdriver-cd/screwdriver/node_modules/sequelize/lib/dialects/mysql/query.js:52:25)
    at /Users/dsagar/Workspace/screwdriver-cd/screwdriver/node_modules/sequelize/lib/sequelize.js:315:28
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async MySQLQueryInterface.select (/Users/dsagar/Workspace/screwdriver-cd/screwdriver/node_modules/sequelize/lib/dialects/abstract/query-interface.js:407:12)
    at async jobs.findAll (/Users/dsagar/Workspace/screwdriver-cd/screwdriver/node_modules/sequelize/lib/model.js:1140:21)
    at async Promise.all (index 2)
    at async exports.Manager.execute (/Users/dsagar/Workspace/screwdriver-cd/screwdriver/node_modules/@hapi/hapi/lib/toolkit.js:60:28)
    at async internals.handler (/Users/dsagar/Workspace/screwdriver-cd/screwdriver/node_modules/@hapi/hapi/lib/handler.js:46:20)
    at async exports.execute (/Users/dsagar/Workspace/screwdriver-cd/screwdriver/node_modules/@hapi/hapi/lib/handler.js:31:20)
    at async Request._lifecycle (/Users/dsagar/Workspace/screwdriver-cd/screwdriver/node_modules/@hapi/hapi/lib/request.js:370:32)
    at async Request._execute (/Users/dsagar/Workspace/screwdriver-cd/screwdriver/node_modules/@hapi/hapi/lib/request.js:280:9)
230817/142820.866, (1692282500866:C02ZM198LVDV:491:llf9ab7u:10002) [response,api,templates,versions] http://192.168.87.128:9001: get /v4/templates/sagar1312%2Fsd-job-template-apple/metrics {"sort":"descending"} 500 (158ms)
```

![image](https://github.com/screwdriver-cd/models/assets/2124590/ce7372a1-99fa-4185-bb91-eabac7a8719e)

nathom committed:
Fix templateId missing issue

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -258,12 +265,18 @@ class TemplateFactory extends BaseFactory {
+            const metricPipelinesConfig = {
+                params: { templateId: templateIds },
+                readOnly: true
+            };
```
This is not an optimal solution. This approach loads the all the fields of job objects from the DB to memory and perform aggregation here.

As we had discussed, we should enhance `datastore-sequelize` to accept input and generate below SQL query so that the aggregation is performed at source (DB)
```
select count(distinct(pipelineId)), templateId
from jobs
group by templateId;
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -258,12 +265,18 @@ class TemplateFactory extends BaseFactory {
 
             const templateIds = templates.map(t => t.id);
 
+            // all jobs with templates
```
```suggestion
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -258,12 +265,18 @@ class TemplateFactory extends BaseFactory {
                 aggregationField: 'templateId'
             };
 
+            const metricPipelinesConfig = {
```
```suggestion
            const jobListConfig = {
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -273,24 +286,35 @@ class TemplateFactory extends BaseFactory {
+                jobFactory.list(metricJobsListConfig),
+                buildFactory.list(metricBuildsListConfig),
+                jobFactory.list(metricPipelinesConfig)
+            ]).then(([jCount, bCount, pCount]) =>
```
```suggestion
            ]).then(([jCount, bCount, jobs]) =>
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -273,24 +286,35 @@ class TemplateFactory extends BaseFactory {
+                    const pipelinesUnique = new Set(pipelineIds);
+                    // The number of pipelines that use this template,
+                    // NOT the number of template uses in pipelines
+                    const pipelineCount = pipelinesUnique.size;
```
```suggestion
                    const pipelineCount = new Set(pipelineIds).size;
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -321,6 +345,63 @@ class TemplateFactory extends BaseFactory {
+                    lastRun: event,
+                    admins: p.admins
+                };
+            })
```
Per discussion, instead of making individual call to EventFactory for each pipeline, let iterate over the pipelines to collect all the eventIds and make just one call to DB

nathom committed:
Only make one call to eventFactory

nathom committed:
Update templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

nathom committed:
Update templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

nathom committed:
Update templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

nathom committed:
Update templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

nathom committed:
Fix lint

nathom committed:
Merge branch 'master' into master

nathom committed:
Fix lint

nathom committed:
Add tests for pipelineUsage

nathom committed:
Merge branch 'master' of https://github.com/nathom/models

nathom committed:
Handle case where template not found

nathom committed:
Fix lint

nathom committed:
Improve tests

Remove comment

Add filter

Fix filter

Update lib/templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

Update test/lib/templateFactory.test.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

Update test/lib/templateFactory.test.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

Update lib/templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

Update lib/templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

Update lib/templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

Update lib/templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

Improve test data

Update lib/templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

Update lib/templateFactory.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

Update test/lib/templateFactory.test.js

Co-authored-by: Dayanand Sagar <sagar1312@gmail.com>

Fix pipelineUsage tests

Remove unnecessary mocks

Add test

Verify calls

Remove only

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -321,6 +341,69 @@ class TemplateFactory extends BaseFactory {
+        const eventFactory = EventFactory.getInstance();
+
+        const pipelineData = await pipelineFactory.list(pipelineConfig);
+        const eventIds = pipelineData.map(p => p.lastEventId).filter(p => p);
```
```suggestion
        const eventIds = pipelineData.filter(p => p.lastEventId !== null).map(p => p.lastEventId);
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -107,6 +107,7 @@ class TemplateFactory extends BaseFactory {
         const [, namespace, name] = TEMPLATE_NAME_REGEX_WITH_NAMESPACE.exec(fullTemplateName);
 
         // Check if template with namespace and name exist, default to using that template
+
```
```suggestion
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -321,6 +341,69 @@ class TemplateFactory extends BaseFactory {
     }
 
+    /**
+     * Get metrics for the pipelines that use the given template
```
```suggestion
     * List of pipelines that use the given template version
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -321,6 +341,69 @@ class TemplateFactory extends BaseFactory {
+     * Get metrics for the pipelines that use the given template
+     * @method getPipelineUsage
+     * @param  {String}     fullTemplateName    Name of the template and the version or tag (e.g. chef/publish@1.2.3)
+     * @return {Promise}                        List of metrics for the pipelines or null if template is not found.
```
```suggestion
     * @return {Promise}                        Resolves to list of pipelines using the template.
```

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1022,6 +1037,53 @@ describe('Template Factory', () => {
+                {
+                    templateId: 2,
+                    pipelineId: 6
+                },
```
Let us nothing for `templateId=2` so that we can test that the pipeline count is `0`
```suggestion        
```

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1062,6 +1130,9 @@ describe('Template Factory', () => {
                             count: 0
+                        },
+                        pipelines: {
+                            count: 1
```
```suggestion
                            count: 0
```

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1022,6 +1037,53 @@ describe('Template Factory', () => {
+            pipelineJobs = [
+                {
+                    templateId: 1,
+                    pipelineId: 4
```
Include `jobId` for all the entries so that we don't confused on why we have duplicate entries of `templateId` and `pipelineId` pair.
```suggestion
                    templateId: 1,
                    pipelineId: 4,
                    jobId: 41
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -321,6 +341,69 @@ class TemplateFactory extends BaseFactory {
+        const template = await this.getTemplate(fullTemplateName);
+
+        if (!template) {
+            return null;
```
Also add a test to cover this scenario
```suggestion
            throw new Error('Template does not exist');
```

nathom replied:
If the template doesn't exist, shouldn't the UI direct to 404, like it does for getTemplate? With this it will return an internal server error. Or would we catch it in `screwdriver`?

sagar1312 replied:
A function should consistently return one type of result for all the scenarios or throw error.
Here, the result should always be list of pipelines. If we return `null`, there could ambiguity on what it represents: no pipelines using the template vs no such template.

Yes, catch it in `screwdriver` and `throw boom.notFound(error)` when the error is `Template does not exist`.

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -321,6 +341,69 @@ class TemplateFactory extends BaseFactory {
+            return null;
+        }
+        const { id } = template;
+        const metricPipelinesConfig = {
```
```suggestion
        const metricJobCountByPipelineListConfig = {
```

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -321,6 +341,69 @@ class TemplateFactory extends BaseFactory {
+        const pipelineData = await pipelineFactory.list(pipelineConfig);
+        const eventIds = pipelineData.filter(p => p.lastEventId !== null).map(p => p.lastEventId);
+        const eventConfig = {
+            params: { id: eventIds }
```
```suggestion
            params: { id: eventIds },
            readOnly: true
```

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1249,6 +1331,164 @@ describe('Template Factory', () => {
+            pipelineFactoryTestOutput = [
+                {
+                    id: 6,
+                    name: 'nathom/sd-uses-template',
```
`name` should be same as `scmRepo.name`. Please correct this for all the entries

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1249,6 +1331,164 @@ describe('Template Factory', () => {
+                        branch: 'main',
+                        name: 'test/repo',
+                        url: 'https://github.com/test/repo/tree/main/pipe1',
+                        rootDir: 'pipe1',
```
Have test data such that combination of `name`, `branch` and `rootDir` is unique

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1249,6 +1331,164 @@ describe('Template Factory', () => {
+                },
+                {
+                    id: 2,
+                    createTime: '2023-08-17T18:18:37.501Z'
```
Have different value for each entry
```suggestion
                    createTime: '2023-07-18T12:18:42.501Z'
```

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1249,6 +1331,164 @@ describe('Template Factory', () => {
+        it('should return empty array when no pipelines are using the template', () => {
+            const expected = [];
+
+            datastore.scan.onCall(0).resolves([]);
```
`jobFactoryMock.list`, `pipelineFactoryMock.list` and `eventFactoryMock.list` are mocked explicitly. Do we need to mock `datastore.scan`? If not please remove it from all the tests
```suggestion
            datastore.scan.onCall(0).resolves([]);
```

nathom replied:
Yes, it's for mocking the initial getTemplate call.

sagar1312 commented on lib/templateFactory.js:
```diff
@@ -321,6 +341,69 @@ class TemplateFactory extends BaseFactory {
+            .list(metricPipelinesConfig)
+            .then(pipelines => pipelines.map(p => p.pipelineId));
+
+        // eslint-disable-next-line global-require
```
If none of the pipelines are using the template, skip rest of the logic. Add test to cover this.
```suggestion
     if(pipelineIds.length === 0) {
       return Promise.resolve([]);
     }

        // eslint-disable-next-line global-require
```

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1249,6 +1331,164 @@ describe('Template Factory', () => {
+            datastore.get.resolves(templateReturnValue);
+            jobFactoryMock.list.resolves([]);
+            pipelineFactoryMock.list.resolves([]);
+            eventFactoryMock.list.resolves([]);
```
```suggestion
```

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1249,6 +1331,164 @@ describe('Template Factory', () => {
+            eventFactoryMock.list.resolves([]);
+
+            return factory.getPipelineUsage(`${namespace}/${name}@1.0.2`).then(pipelines => {
+                assert.deepEqual(pipelines, expected);
```
Also verify that `datastore.get` and `jobFactoryMock.list` was called with the right arguments

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1249,6 +1331,164 @@ describe('Template Factory', () => {
+        it('should list pipelines using template version', () => {
+            const expected = returnValue;
+
+            datastore.scan.onCall(0).resolves([templateReturnValue]);
```
```suggestion
```

nathom replied:
This is necessary.

sagar1312 commented on test/lib/templateFactory.test.js:
```diff
@@ -1249,6 +1331,164 @@ describe('Template Factory', () => {
+            eventFactoryMock.list.resolves(eventFactoryTestOutput);
+
+            return factory.getPipelineUsage(`${namespace}/${name}@1.0.2`).then(pipelines => {
+                assert.deepEqual(pipelines, expected);
```
Same here. Also verify that all the mocked methods were called with the right arguments.
